samples/matter/common/src/ota_multi_image_downloader.cpp

```diff
@@ -0,0 +1,152 @@
+
+int OTAMultiImageDownloader::Init(const device *flash, ssize_t size)
+{
+	if (!device_is_ready(flash)) {
+		ChipLogError(SoftwareUpdate, "Flash device is not ready!");
+		return -EFAULT;
+	}
+	return Init(flash, mOffset, size);
+}
+
+int OTAMultiImageDownloader::Init(const device *flash, size_t offset, size_t size)
+{
+	mBuffer = (uint8_t *)Platform::MemoryAlloc(kBufferSize);
+	return stream_flash_init(&mStream, flash, mBuffer, kBufferSize, offset, size, nullptr);
+}
+
+int OTAMultiImageDownloader::Apply(const device *flash)
+{
+	// Reinit stream with different flash partition
+	int err = Init(flash, mTargetOffset, mTargetImgSize);
+	if (err != 0) {
+		ChipLogError(SoftwareUpdate, "Cannot reinitialize flash stream");
+		return -EBUSY;
+	}
+
+	Platform::ScopedMemoryBuffer<uint8_t> buffer;
+	if (!buffer.Alloc(kBufferSize)) {
+		return -ENOMEM;
+	}
+
+	for (size_t offset = 0; offset < mImgSize; offset += kBufferSize) {
+		err = flash_read(flash, mOffset + offset, buffer.Get(), kBufferSize);
+		if (err != 0) {
+			ChipLogError(SoftwareUpdate, "Failed to read data chunk from flash");
+			return -EIO;
+		}
+		err = Write(buffer.Get(), kBufferSize);
+		if (err != 0) {
+			ChipLogError(SoftwareUpdate, "Failed to write data chunk to flash");
+			return -EIO;
+		}
+	}
+	return Finalize();
+}
+
+int OTAMultiImageDownloader::Write(const uint8_t *chunk, size_t chunkSize)
+{
+	return stream_flash_buffered_write(&mStream, chunk, chunkSize, false);
+}
+
+int OTAMultiImageDownloader::Finalize()
+{
+	int err = stream_flash_buffered_write(&mStream, nullptr, 0, true);
+	Platform::MemoryFree(mBuffer);
+	return err;
+}
+
+// Implementation of stateless OTAMultiImageDownloaders class
+
+int OTAMultiImageDownloaders::Open(int id, size_t size)
+{
+	int dfuTargetId = id < 2 ? id : id - 2; // dfu_target only accepts fixed 0 and 1 slots in secondary
+						// partition
+	if (DfuTargetActionNeeded(id)) {
+		int err = dfu_target_init(DFU_TARGET_IMAGE_TYPE_MCUBOOT, dfuTargetId, size, nullptr);
+		if (err) {
+			ChipLogError(SoftwareUpdate, "Cannot initialize DFU target: %d", err);
+			return err;
+		}
+	}
+
+	OTAMultiImageDownloaders::SetCurrentImageId(id);
+	return OTAMultiImageDownloaders::GetCurrentImage()->Init(sExtFlash, size);
+}
+
+int OTAMultiImageDownloaders::Write(const uint8_t *chunk, size_t chunk_size)
+{
+	if (DfuTargetActionNeeded(OTAMultiImageDownloaders::CurrentImageId())) {
+		int err = dfu_target_write(chunk, chunk_size);
+		if (err < 0) {
+			ChipLogError(SoftwareUpdate, "Cannot write DFU target: %d", err);
+			return err;
+		}
+	}
+
+	return OTAMultiImageDownloaders::GetCurrentImage()->Write(chunk, chunk_size);
+}
+
+int OTAMultiImageDownloaders::Close(bool success)
+{
+	if (DfuTargetActionNeeded(OTAMultiImageDownloaders::CurrentImageId())) {
+		int err = success ? dfu_target_done(success) : dfu_target_reset();
+		if (err) {
+			ChipLogError(SoftwareUpdate, "Cannot close DFU target: %d", err);
+			return err;
```

Comment:
If we exit here, `Finalize()` will not be called and the memory buffer will leak (given that `Initialize` assumes the buffer has been released on the previous usage).

Reply:
Added `Cleanup` method.